### PR TITLE
ci: enhance build process

### PR DESCRIPTION
**Pull request by georglauterbach**
# Description

Enhances `Dockerfile` by

1. outsourcing packages installation
2. using `COPY --link` for ClamAV anti-virus bytecode

I am not sure whether using `DOCKER_BUILDKIT` is a breaking change. Technically not, because the build process is not taken into account for SemVer, right?

PS: Sorry for the noise in the commit message. Accidentally commit the wrong files first with wrong description.
PPS: Note, this reverts #1965 implicitly. That is okay, since Debian 11 now ships with the latest version of Fail2Ban

## Type of change

- [x] Improvement (non-breaking change that does improve existing functionality)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (README.md or the documentation under `docs/`)
- [x] If necessary I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: target/scripts/helpers/log.sh
Comment (georglauterbach) [inline target/scripts/helpers/log.sh:81]: Side note: had to adjust this here because, as it turns out, `LGRAY` is never set, and the `set -u` will take care of exiting with error during build in `packages.sh`.

File: Dockerfile
Comment (polarathene) [inline Dockerfile:29]: Have you looked into if this invalidates cache when a change is made in `packages.sh` for a new build?

This is not an approach I recall seeing elsewhere when that matters. And Dockers ["Best Practices" guide for build cache](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#leverage-build-cache) states:

> Aside from the `ADD` and `COPY` commands, cache checking does not look at the files in the container to determine a cache match.
> For example, when processing a `RUN apt-get -y update` command the files updated in the container are not examined to determine if a cache hit exists.
> In that case just **the command string itself is used to find a match**.

I think this isn't an issue though since we're using an earlier `COPY` for the script, which will perform an invalidation AFAIK.

---

Due to the dependence on `log.sh`, any changes to that is going to invalidate this layer as well which is a shame, but hopefully rarely an issue going forward.

File: target/scripts/build/packages.sh
Comment (polarathene) [inline target/scripts/build/packages.sh:110]: I'd likewise be curious about the impact of each of these if you have time to inspect with `dive`.

I wasn't sure if looking over `apt-get` docs if there was any reason to use both `clean` and `autoclean` for example.

**Review by polarathene — CHANGES_REQUESTED** [review]
LGTM :+1: 

Just a few questions in the feedback, and some optional improvements.

The only change that I'd like to see is it looks like you might have forgotten to remove the `FAIL2BAN_*` ARGs that https://github.com/docker-mailserver/docker-mailserver/pull/1965 introduced?

All packages are accounted for still which is appreciated before dropping them in a follow-up :+1:

File: target/scripts/build/packages.sh
Comment (polarathene) [inline target/scripts/build/packages.sh:37]:
```suggestion
  _log 'warn' 'Applying workaround for Postfix bug (see https://github.com/docker-mailserver/docker-mailserver/issues/2023#issuecomment-855326403)'
```

File: Dockerfile
Comment (polarathene) [reply Dockerfile:29]: There is [one way around the issue with cache invalidation](https://docs.docker.com/engine/reference/builder/#run---mounttypecache), which local builds could at least take advantage of. I don't think it'd affect our CI caching negatively either.

Just for reference, as it's mostly only going to matter for local builds if you're frequently adjusting `packages.sh` or `log.sh`. The linked docs provide an example specifically with `apt` usage:

```Dockerfile
# syntax=docker/dockerfile:1
FROM ubuntu
RUN rm -f /etc/apt/apt.conf.d/docker-clean; echo 'Binary::apt::APT::Keep-Downloaded-Packages "true";' > /etc/apt/apt.conf.d/keep-cache
RUN --mount=type=cache,target=/var/cache/apt,sharing=locked \
  --mount=type=cache,target=/var/lib/apt,sharing=locked \
  apt update && apt-get --no-install-recommends install -y gcc
```

> _`RUN --mount` allows you to create mounts that process running as part of the build can access._
> _This can be used to bind files from other part of the build without copying, accessing build secrets or ssh-agent sockets, or creating cache locations to speed up your build._
> 
> _`RUN --mount=type=cache`_
> _This mount type allows the build container to cache directories for compilers and package managers._

No need to look into that for this PR, just making a note of it here.

That'd change how we're handling post-install cleanup I think, so isn't appropriate for this PR where we want to keep large diffs of shuffling lines around easy to track before further revision.

I'm not sure where it mounts that locally, presumably a docker data / volume container instead of a bind volume mount.

File: target/scripts/build/packages.sh
Comment (polarathene) [reply target/scripts/build/packages.sh:43]: Just another note. Not suitable for this PR.

BuildKit seems to allow specifying a hostname which might work via the build ENV [`BUILDKIT_SANDBOX_HOSTNAME`](https://docs.docker.com/engine/reference/builder/#buildkit-built-in-build-args). Might be a little cleaner as a workaround, but not sure if it is any better. At least with the current approach the workaround and problem package are nicely co-located together.

File: target/scripts/build/packages.sh
Comment (georglauterbach) [reply target/scripts/build/packages.sh:30]: >     * `apt-transport-https`, `ca-certificates`, `curl` were added later with the other packages. Is there a benefit to adding them here? (_no issue with that, just curious about the change_)

This is done in preparation for upcoming PRs that require these packages early (e.g. adding PPAs).

>     * `gnupg2`, `ssl-cert` are newly added via this PR. What are these needed for? `rspamd`? If so they should be deferred until that?

Indeed, `gnupg2` is for adding the PPA of `rspamd`. When I reviewed the installation log, I came across a few warnings from packages that want `ssl-cert` to be installed, so I put it here too.

I disagree that they should be deferred - this is more work for me without adding any benefit really, from my PoV. These packages are more or less essential and I think having them available early in the process is beneficial, installing `rspamd` or not.

>     * It doesn't look like we used `update-ca-certificates` previously, and while the `update` is done (_a little above the highlighted range_), you've also swapped `dist-upgrade` for `upgrade`?

The `update-ca-certificate` is done now because I saw some warnings in the logs about CA certificates, which are solved by updating (sorry, can't remember exactly what the message was, but `c_rehash` logged the warning).

> Is there a reason for preferring to switch from `dist-upgrade` to `upgrade`?

This was a lint from Hadolint about `dist-upgrade`, and the [AskUbuntu forum](https://askubuntu.com/a/226213) explains that `dist-upgrade` may remove packages. This is undesired, and Hadolint is right, `upgrade` is preferred I guess.

File: target/scripts/build/packages.sh
Comment (georglauterbach) [reply target/scripts/build/packages.sh:6]: You're welcome :)

File: Dockerfile
Comment (georglauterbach) [reply Dockerfile:48]: Hmm, I am not sure. I think the command is pretty self-explanatory, but on the other hand, I don't mind either. While the rate-limiting indeed was a problem, I think listing problems we solve in comments in the whole code base would increase code base size by 9000% 😆

File: target/scripts/build/packages.sh
Comment (georglauterbach) [reply target/scripts/build/packages.sh:11]: Yes, this will just increase log verbosity. We need the `-y` flag I think because the whole process is non-interactive (not sure whether `DEBIAN_FRONTEND=noninteractive` is enough). I will test and adjust.

File: target/scripts/build/packages.sh
Comment (georglauterbach) [reply target/scripts/build/packages.sh:43]: Actually, I think this is suitable for this PR - after all, this PR is about the build process. I will commit your suggestion :)

I have looked at `[BUILDKIT_SANDBOX_HOSTNAME](https://docs.docker.com/engine/reference/builder/#buildkit-built-in-build-args)` as well. I think our current solution is better since we have the workaround in one place, as you said.

---

> It does work, but that ARG requires the Dockerfile adds `# syntax=docker/dockerfile:1`

`# syntax=docker/dockerfile:1` is already added by this PR :)

File: Dockerfile
Comment (georglauterbach) [reply Dockerfile:29]: The `COPY` will indeed invalidate cache, so the `RUN` is executed again. And `log.sh` is rarely touched, so this is a non-issue if you ask me.

---

👍🏼 when it comes to the solution about the APT package cache. Might do another PR in the future for that.

File: target/scripts/build/packages.sh
Comment (georglauterbach) [reply target/scripts/build/packages.sh:106]: I could do that :) but it is a hassle and I think it does not do harm 😆

File: target/scripts/build/packages.sh
Comment (georglauterbach) [reply target/scripts/build/packages.sh:110]: I guess `autoclean` is not needed when `clean` is run. I found [this](https://askubuntu.com/questions/3167/what-is-difference-between-the-options-autoclean-autoremove-and-clean), and I guess removing `autoclean` is is a good idea.

Comment (georglauterbach) [reply target/scripts/build/packages.sh]: It turns out that `c_rehash` is not required either.. I will remove this too.

**Comment by georglauterbach** [comment]
I removed the left-oever Fail2Ban `ARG`s from `Dockerfile` too :)

File: Dockerfile
Comment (polarathene) [reply Dockerfile:43]: > I think listing problems we solve in comments in the whole code base would increase code base size by 9000%

I don't like it either sometimes, but history is messy with all the refactoring, and I've seen some past contributions make assumptions because they weren't aware of why something was done :\

When refactoring settles down, we can probably offload plenty of it, especially if we introduce some technical maintainer doc pages.

For now a small comment is much easier to get context about a decision than wading through PRs I review verbosely :stuck_out_tongue:

File: target/scripts/build/packages.sh
Comment (polarathene) [reply target/scripts/build/packages.sh]: > I disagree that they should be deferred - this is more work for me without adding any benefit really, from my PoV.

I understand that, and I know it's annoying to ask.

It's mostly for traceability in future. Many times if I attempt to wade through history searching for a commit message or PR that explains why something was added, even with packages it's not always evident beyond possible context of what the commit/PR is contributing.

In this case your comment at least adds some clarification if someone will search through the PR review process here. But it's not really related to `ci: enhance build process` and is very much belonging to your PR that will introduce `rspamd`.

Likewise, if someone goes through history for what was added to bring in `rspamd` support, they'd lack the context of these packages being associated to that, as the PR adding `rspamd` would presumably not even mention the relation.

---

Realistically... we've gained no new long-term maintainers since I got involved, and who knows when/if a new one will come along. I'm not going to block approval over this :+1: 

---

> Turns out, `update-ca-certificate` is not actually needed. I was unable to reproduce a warning message. I will remove it again.

I'm not against that btw, although I would assume it's part of the `upgrade` process when certs are updated, but I could be mistaken.

`c_rehash` was related to that I think. I tried to look back through `git blame` for where and why it was added, but the commit that introduced it was the "Final Migration" one with an invalid PR reference :laughing: 

---

> This was a lint from Hadolint about `dist-upgrade`, and the [AskUbuntu forum](https://askubuntu.com/a/226213) explains that `dist-upgrade` may remove packages. This is undesired, and Hadolint is right, `upgrade` is preferred I guess.

:+1:

**Review by polarathene — DISMISSED** [review]
:tada:

File: target/scripts/build/packages.sh
Comment (georglauterbach) [reply target/scripts/build/packages.sh]: Alright, I can make an effort here :) I restored the original order in cde762061f9d08c54733583c57a94b8076bc8ef9 :)

File: target/scripts/build/packages.sh
Comment (casperklein) [inline target/scripts/build/packages.sh:11]: Regex removed and simplified:
```suggestion
[[ ${LOG_LEVEL} == trace ]] && QUIET='-y' || QUIET='-qq'
```

**Review by casperklein — COMMENTED** [review]
> Moreover, this commit implicitly removed the explicit Fail2Ban
> installation and therefore #1965, BECAUSE Debian 11 now ships with the
> latest and greatest Fail2Ban version

I would like to stick with the official debian package from fail2ban. This way we ensure, to get the latest version (also in the future, when Debian might stick with an older version as happen in the past).

**Comment by polarathene** [comment]
> I would like to stick with the official debian package from fail2ban

Do you mind if we bring that back in a follow-up PR that is introduced as a separate stage?

> While we at it, don't we want to use the most recent dovecot version too?

Likewise, I don't mind this if we keep everything required to support it as a separate stage.

This way the changes between switching from the Debian package manager to pulling in external packages to support has cleaner history of changes.

Eventually, I'd like to evaluate using/supporting a different base image such as Alpine, so keeping such changes isolated is helpful.

File: target/scripts/build/packages.sh
Comment (polarathene) [reply target/scripts/build/packages.sh]: > Alright, I can make an effort here :)

I'm not sure what you're referring to with that?

I was requesting that `gnupg2` and `ssl-cert` be deferred to the relevant `rspamd` PR that would likely follow shortly after this is merged, or after the 11.2 release.

I just see them blended into the package list. Re-locating them wasn't of interest to me. I wanted them out of this PR, but did not want to mandate it. This is not a request to revert, I'll approve regardless of what you choose to do here, just a preference to have the packages belong in a PR they're intended for.

> I restored the original order in [cde7620](https://github.com/docker-mailserver/docker-mailserver/commit/cde762061f9d08c54733583c57a94b8076bc8ef9) :)

For `apt-transport-https`, `ca-certificates`, `curl` - I had said I did not mind these being re-located, I was just noting the observation as part of my review for tracking the shifting of packages in this PR :sweat_smile:

**Review by polarathene — APPROVED** [review]

File: target/scripts/build/packages.sh
Comment (georglauterbach) [reply target/scripts/build/packages.sh:11]: I would likewise do follow-up here, since the log-level-querying is now on `master` and this can be adjusted a bit. I hope this is okay:)

Good catch though!

**Comment by georglauterbach** [comment]
> I would like to stick with the official debian package from fail2ban. This way we ensure, to get the latest version (also in the future, when Debian might stick with an older version as happen in the past).

I don't quiet get that. With this PR, we're basically using `apt-get install fail2ban` - this is what I understand as "the official Debian package". Could you elaborate, please?:)

**Comment by polarathene** [comment]
> this is what I understand as "the official Debian package". Could you elaborate, please?:)

I think he was referring to the [`.deb` release from the official Github releases](https://github.com/fail2ban/fail2ban/releases/tag/0.11.2)? (aka `FAIL2BAN_DEB_URL` that was used to pull the release).

There's not been a new release for sometime since though, and the next Debian release (expected around mid 2023?) may not get it (_depending on if a release happens before the cutoff point for Debian_). There is over 350 commits of updates since 11.2 release, so if that happens we may have to install from Github releases again, instead of just the [`fail2ban` package](https://packages.debian.org/bullseye/fail2ban) (_currently the latest 11.2 release_).

I'm not fussed either way, but if we don't use the package, it should be contained in it's own stage. @casperklein could contribute that after this PR is merged?

**Comment by georglauterbach** [comment]
CI seems to be working as expected with #2753 🎉

File: target/scripts/build/packages.sh
Comment (georglauterbach) [reply target/scripts/build/packages.sh]: Done in c6aa12436cbd0fc43f8dfac45d482f1e36689234

File: target/scripts/build/packages.sh
Comment (georglauterbach) [reply target/scripts/build/packages.sh]: Done in c6aa12436cbd0fc43f8dfac45d482f1e36689234

**Comment by georglauterbach** [comment]
I adjusted the PR according to PR feedback :) @casperklein I hope you don't mind the Fail2Ban package change - I agree with @polarathene that F2B could get it's own stage in a future PR :)

**Review by polarathene — APPROVED** [review]

**Comment by casperklein** [comment]
> I think he was referring to the [.deb release from the official Github releases](https://github.com/fail2ban/fail2ban/releases/tag/0.11.2)? (aka FAIL2BAN_DEB_URL that was used to pull the release).

Correct. Given, that we almost already know, that we'll switch back to the .deb file sooner or later, why removing it now? Just to generate more work?

> Do you mind if we bring that back in a follow-up PR that is introduced as a separate stage?

Just to be clear, the separate stage would just be for downloading the .deb file?

```
FROM stage-base as stage-fail2ban

ARG FAIL2BAN_DEB_URL=https://github.com/fail2ban/fail2ban/releases/download/0.11.2/fail2ban_0.11.2-1.upstream1_all.deb
ARG FAIL2BAN_DEB_ASC_URL=${FAIL2BAN_DEB_URL}.asc
ARG FAIL2BAN_GPG_PUBLIC_KEY_ID=0x683BF1BEBD0A882C
ARG FAIL2BAN_GPG_PUBLIC_KEY_SERVER=hkps://keyserver.ubuntu.com
ARG FAIL2BAN_GPG_FINGERPRINT="8738 559E 26F6 71DF 9E2C  6D9E 683B F1BE BD0A 882C"

gpg --keyserver ${FAIL2BAN_GPG_PUBLIC_KEY_SERVER} \
    --recv-keys ${FAIL2BAN_GPG_PUBLIC_KEY_ID} 2>&1 && \
  curl -Lkso fail2ban.deb ${FAIL2BAN_DEB_URL} && \
  curl -Lkso fail2ban.deb.asc ${FAIL2BAN_DEB_ASC_URL} && \
  FINGERPRINT=$(LANG=C gpg --verify \
  fail2ban.deb.asc fail2ban.deb 2>&1 \
    | sed -n 's#Primary key fingerprint: \(.*\)#\1#p') && \
  if [[ -z ${FINGERPRINT} ]]; then \
    echo "ERROR: Invalid GPG signature!" >&2; exit 1; fi && \
  if [[ ${FINGERPRINT} != "${FAIL2BAN_GPG_FINGERPRINT}" ]]; then \
    echo "ERROR: Wrong GPG fingerprint!" >&2; exit 1; fi
```

```
COPY --from=stage-fail2ban /fail2ban.deb /fail2ban.deb
RUN dpkg -i /fail2ban.deb && rm /fail2ban.deb
```

Serious question: Is this really worth it? IMO, this should be in the new introduced packages.sh and handled there.

**Comment by georglauterbach** [comment]
I have reviewed the version in the Debian 11 container, and it's the very latest we are using as well. I do not see, at the moment, why we should have this extra code that was placed there to get major version 11 (for `nftables` support I think) when the default Debian package is as good now.

If we, however, decide to bring our own extra code back, I agree with @casperklein that this should go into `packages.sh`, if there are no good reasons for doing it in the Dockerfile. But, I would actually first go with removing the extgra code for now (and hopefully forever), since we do not need the very latest of the latest now, do we?

**Comment by polarathene** [comment]
**TL;DR:** I don't have a strong enough opinion about the `fail2ban` decision. I'd rather see this PR get resolved, whatever makes that easier to happen is fine by me.

---

> Serious question: Is this really worth it? IMO, this should be in the new introduced packages.sh and handled there.

I was preferring a separate stage because it interested me in building an image without fail2ban, or being able to change the base image to something like alpine where `.deb` would not be supported.

This would apply for pretty much anything in `packages.sh` though, and I don't have a strong need for `fail2ban` being isolated (it'd be more work beyond this since we have other support in scripts related to it, and tbh I doubt I'll find the time).

---

I'm fine with it being maintained in `packages.sh` as it's own function call :+1: 

`Dockerfile` is also fine, but it seems like a bunch of extra noise for something that is just a package. Upstream has had many commits, but I haven't seen a release in a while.

I don't know when a new release of f2b will happen or how important that is to the users of our image.

We could do the same for `dovecot` and `postfix` and whatever other packages too, but how many users need it, and if they do is Debian the right choice for a base image, or is something that updates packages more frequently like Alpine better? (_I'm sure it has it's own share of problems too_)

---

As a maintainer, what I care most about (_and for any future maintainers if we're lucky_), is that we have something well organized and scoped.

I'd like to say it'd be a `git blame` away, but we all know how frustrating that can be to track down the original PR for something to get more context on why the project chose to do something, and what in the project is related to a feature/change (@georglauterbach _for example originally wanted to mix in changes for `rspamd` into this PR, which would contribute to the problem_)

**Review by casperklein — APPROVED** [review]
Alright. I am still not 100% convinced, but I don't want to be the blocker here.

I bet, that the time will come when Debian again will offer not the latest version of f2b. Especially with security related software, I prefer the latest version with it's features.
"But then we can add it again".. Yeah. But that's just extra work in the future while now it would be an ease to keep the current f2b solution.

Dovecot/F2B are both very easy to keep up2date, as they offer .deb files or a repository. For Postfix unfortunately, there isn't such a option.

**Comment by georglauterbach** [comment]
Okay, alright :) 

Next week, I will tackle the cache issue in our CI and take care of the Fail2Ban concern raised here.

**Comment by casperklein** [comment]
Not sure if I understand you correctly. Are you going to adjust this PR or merge it soon? If the later one, I'll provide a follow up with f2b included in `packages.sh` and optional if there is consensus also with [Dovecot](https://repo.dovecot.org/).

**Comment by georglauterbach** [comment]
> Not sure if I understand you correctly. Are you going to adjust this PR or merge it soon? If the later one, I'll provide a follow up with f2b included in `packages.sh` and optional if there is consensus also with [Dovecot](https://repo.dovecot.org/).

I was about to do the former, but if you want to do the latter, I won't hinder you. I will need to resolve conflicts in either way. Adding the Dovecot package via .deb files seems like a good idea, so maybe a follow-up is a good idea after all?

**Comment by polarathene** [comment]
> I'll provide a follow up with f2b included in `packages.sh` and optional if there is consensus also with [Dovecot](https://repo.dovecot.org/).

:+1: for adding Dovecot. I would appreciate if both f2b and dovecot live in their own separate function calls in `packages.sh` though if not too much trouble. Just a preference though, approach it however you like :)

**Comment by georglauterbach** [comment]
@casperklein which way do you prefer? Shall I integrate the changes into this PR or shall I or will you do a follow up PR?

**Comment by casperklein** [comment]
Follow up :+1:

**Comment by georglauterbach** [comment]
> Follow up :+1:

👍 Then I will provide a follow up. I will resolve conflicts for this PR on Monday and then we can merge this :)

**Comment by casperklein** [comment]
I meant, I'll provide a follow up. I've already something prepared 😉

**Comment by georglauterbach** [comment]
> I meant, I'll provide a follow up. I've already something prepared 😉 

Very nice 🚀

**Comment by georglauterbach** [comment]
I resolved the merge conflicts and updated the docker/build-push action so that the version used across our repository is the same everywhere.

**Review by casperklein — APPROVED** [review]

**Review by polarathene — APPROVED** [review]

**Comment by casperklein** [comment]
> There's not been a new release for sometime since though

Boom, there it [is](https://github.com/fail2ban/fail2ban/releases/tag/1.0.1) 🚀

**Comment by polarathene** [comment]
> I bet, that the time will come when Debian again will offer not the latest version of f2b. Especially with security related software, I prefer the latest version with it's features.

Ok, I didn't expect the release to happen that soon after making such a statement :laughing: :sweat_smile: 

That just leaves `postsrsd` (_which has been fixed quickly after reporting the bug_) for the ulimit issue.

**Pull request merged**
